File: ESTRUCTURA DE DATOS/DEBERES/SEGUNDO_PARCIAL/DEBER2/PROYECTO_NRC1978/LISTA_SIMPLE_PYTHON/PYTHON/validaciones.py
import re
import msvcrt  # Usado para simular _getch() en Windows
import sys

class Validaciones:
    def ingresar_string(self, msj):
        input_str = ""
        print(msj, end="", flush=True)

        while True:
            c = msvcrt.getch().decode("utf-8")  # Lee un solo carácter

            if c.isalpha():  # Si el carácter es alfabético
                input_str += c
                print(c, end="", flush=True)
            elif c == '\b' and input_str:  # Si es una retroceso y hay caracteres
                input_str = input_str[:-1]
                print("\b \b", end="", flush=True)
            elif c == '\r':  # Enter (retorno de carro)
                if input_str:
                    break
                else:
                    print("\a", end="", flush=True)  # Beep si no hay entrada
            else:
                print("\a", end="", flush=True)  # Beep para otros caracteres

        print()
        return input_str

    def ingresar_cedula(self, msj):
        cedula = ""
        print(msj, end="", flush=True)

        while True:
            c = msvcrt.getch().decode("utf-8")  # Lee un solo carácter

            if c.isdigit():  # Si el carácter es numérico
                cedula += c
                print(c, end="", flush=True)
            elif c == '\b' and cedula:  # Si es una retroceso y hay caracteres
                cedula = cedula[:-1]
                print("\b \b", end="", flush=True)
            elif c == '\r':  # Enter (retorno de carro)
                if cedula:
                    break
                else:
                    print("\a", end="", flush=True)  # Beep si no hay entrada
            else:
                print("\a", end="", flush=True)  # Beep para otros caracteres

        print()
        return cedula

    def ingresar_correo(self, msj):
        correo = input(msj)
        return correo

    def validar_cedula(self, cedula):
        if len(cedula) != 10:
            return False

        suma = 0
        for i in range(9):
            digito = int(cedula[i])
            if i % 2 == 0:
                digito *= 2
                if digito > 9:
                    digito -= 9
            suma += digito

        ultimo_digito = int(cedula[9])
        digito_verificador = (10 - (suma % 10)) % 10

        return ultimo_digito == digito_verificador

    def validar_correo(self, correo):
        pattern = r"(\w+)(\.{1}\w+)*@(\w+)(\.{1}\w+)+"
        return re.match(pattern, correo) is not None

File: ESTRUCTURA DE DATOS/PROYECTO_SEGUNDO_PARCIAL/CODIGO - ARBOLES/validaciones.cpp
#include "Validaciones.h"  
#include <iostream>
#include <conio.h>
#include <regex>

string Validaciones::ingresarString(const string &msj)
{
    string input;
    char c;

    cout << msj;

    while (true)
    {
        c = getch();  

        if (isalpha(c))
        {
            input += c;
            cout << c;
        }
        else if (c == '\b' && !input.empty())  
        {
            input.pop_back();
            cout << "\b \b";
        }
        else if (c == '\r')  
        {
            if (!input.empty())  
            {
                break;
            }
            else
            {
                cout << "\a";  
            }
        }
        else
        {
            cout << "\a";  
        }
    }

    cout << endl;
    return input;
}

string Validaciones::ingresarCedula(const string &msj)
{
    string cedula;
    char c;

    cout << msj;

    while (true)
    {
        c = _getch();  

        if (isdigit(c))
        {
            cedula += c;
            cout << c;
        }
        else if (c == '\b' && !cedula.empty())  
        {
            cedula.pop_back();
            cout << "\b \b";
        }
        else if (c == '\r')  
        {
            if (!cedula.empty())  
            {
                break;
            }
            else
            {
                cout << "\a";  
            }
        }
        else
        {
            cout << "\a";  
        }
    }

    cout << endl;
    return cedula;
}

string Validaciones::ingresarCorreo(const string &msj)
{
    string correo;
    cout << msj;
    cin >> correo;
    return correo;
}

bool Validaciones::validarCedula(const string &cedula)
{
    if (cedula.length() != 10)
    {
        return false;
    }

    int suma = 0;
    for (int i = 0; i < 9; i++)
    {
        int digito = cedula[i] - '0';
        if (i % 2 == 0)
        {
            digito *= 2;
            if (digito > 9)
            {
                digito -= 9;
            }
        }
        suma += digito;
    }

    int ultimoDigito = cedula[9] - '0';
    int digitoVerificador = (10 - (suma % 10)) % 10;

    return ultimoDigito == digitoVerificador;
}

bool Validaciones::validarCorreo(const string &correo)
{
    const regex pattern(R"((\w+)(\.{1}\w+)*@(\w+)(\.{1}\w+)+)");
    return regex_match(correo, pattern);
}